src/jit/assertionprop.cpp

```diff
@@ -848,43 +864,93 @@ Compiler::AssertionIndex Compiler::optCreateAssertion(GenTreePtr op1, GenTreePtr
         //
         // Set op1 to the instance pointer of the indirection
         // 
-        if ((op1->gtOper == GT_ADD) && (op1->gtType == TYP_BYREF))
-        {
-            op1 = op1->gtOp.gtOp1;
 
-            if ((op1->gtOper == GT_ADD) && (op1->gtType == TYP_BYREF))
+        ssize_t offset = 0;
+        while ((op1->gtOper == GT_ADD) && (op1->gtType == TYP_BYREF))
+        {
+            if (op1->gtGetOp2()->IsCnsIntOrI())
+            {
+                offset += op1->gtGetOp2()->gtIntCon.gtIconVal;
+                op1 = op1->gtGetOp1();
+            }
+            else if (op1->gtGetOp1()->IsCnsIntOrI())
             {
-                op1 = op1->gtOp.gtOp1;
+                offset += op1->gtGetOp1()->gtIntCon.gtIconVal;
+                op1 = op1->gtGetOp2();
+            }
+            else
+            {
+                break;
             }
         }
 
-        if (op1->gtOper != GT_LCL_VAR)
+        if (fgIsBigOffset(offset) || op1->gtOper != GT_LCL_VAR)
         {
             goto DONE_ASSERTION;  // Don't make an assertion
         }
 
         unsigned lclNum = op1->gtLclVarCommon.gtLclNum;    noway_assert(lclNum  < lvaCount);
         LclVarDsc * lclVar = &lvaTable[lclNum];
 
+        ValueNum vn;
+
         //
         // We only perform null-checks on GC refs 
         // so only make non-null assertions about GC refs
```

Comment:
You are no making assertions when you have a TYP_BYREF (and we are not in Local AssertionProp), so this comment needs updating.